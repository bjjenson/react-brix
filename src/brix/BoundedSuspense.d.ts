export interface ISuspense {
  fallback: React.ReactType<React.ReactPropTypes>
  boundary: React.ReactType<React.ReactPropTypes>
}

declare const BoundedSuspense: React.ComponentType<ISuspense>
export default BoundedSuspense
