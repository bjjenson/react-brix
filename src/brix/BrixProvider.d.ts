import * as React from 'react'
import { Map } from 'immutable'

export interface BrixProps extends React.HTMLAttributes<HTMLDivElement> {
  value: any,
}

export type BrixProvider = React.ComponentType<BrixProps>
export function getBrixContext(): React.Context<{ state: Map<string, any> }>
export function getBrixConsumer(): React.Consumer<BrixProps>
