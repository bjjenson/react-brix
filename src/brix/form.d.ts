import { Map } from 'immutable'

export type FieldTypes =
  'text' |
  'boolean' |
  'select' |
  'number'

export interface IFieldValues {
  [key: String]: any
}

export interface IFormFieldArgs<T = String> {
  name: String
  type: FieldTypes
  helperText: String
  optional: Boolean
  requiredMessage?: String
  label: String
  normalize?: Function
  value: T
  validate?: (value: T, fieldName: String, getValues: () => IFieldValues) => String
  options?: Array<ISelectOptions>
  valueFromTarget?: (target: Object) => T
}

export interface ISelectOptions {
  value: String
  label: String
}

export interface IFormTextFieldProps {
  error: Boolean
  helperText: String
  label: String
  value: String
  // Handlers
  onBlur: Function
  onChange: Function
}

export interface IFormField {
  props: IFormTextFieldProps
  setValidationResult: (result: String) => void
  setValue: (value: String) => void
  validate: () => void
}

export interface IForm {
  [key: String]: IFormTextFieldProps
  Form: React.Component
  submit: Function
  addField: (field: IFormFieldArgs) => void
  removeField: (fieldName: String) => void
}

export interface IFormErrors {
  [key: String]: String
}

export interface IFormValues {
  [key: String]: any
}

export interface IFormProps {
  fields: Array<IFormFieldArgs>
  setValue: (fieldName: String, value: any) => void
  submit: (values: IFormValues) => Promise
  validate?: (values: IFormValues) => IFormErrors
  initialValues?: Map<String, any>
}

export function useForm(props: IFormProps): IForm
