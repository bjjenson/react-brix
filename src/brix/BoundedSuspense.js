import React from 'react'
import PropTypes from 'prop-types'

class BoundedSuspense extends React.Component {
  state = { hasError: false }

  static getDerivedStateFromError(error) {
    return { hasError: true }
  }

  componentDidCatch(error, info) {
    console.log(error, info)
  }

  render() {
    const { fallback, boundary, children } = this.props
    const { hasError } = this.state
    if (hasError) {
      return boundary
    }

    return (
      <React.Suspense fallback={fallback}>
        {children}
      </React.Suspense>
    )
  }
}

BoundedSuspense.propTypes = {
  fallback: PropTypes.node.isRequired,
  boundary: PropTypes.node,
}

BoundedSuspense.defaultProps = {
  boundary: <h3 style={{ color: 'red' }}>Something is bad</h3>,
}

export default BoundedSuspense