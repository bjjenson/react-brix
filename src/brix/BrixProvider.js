import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { Map } from 'immutable'

const BrixContext = React.createContext()

export const BrixProvider = ({ value, children }) => {
  const [state, setState] = useState(value)

  return (
    <BrixContext.Provider value={{ state, setState }}>
      {children}
    </BrixContext.Provider>
  )
}

BrixProvider.propTypes = {
  value: PropTypes.object,
}

BrixProvider.defaultProps = {
  value: Map(),
}

export function getBrixContext() {
  return BrixContext
}

export function getBrixConsumer() {
  return BrixContext.Consumer
}
