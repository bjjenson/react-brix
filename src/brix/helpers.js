import { fromJS } from 'immutable'

export const coerceDataToImmutable = data =>
  isPlainObject(data) || Array.isArray(data) ? fromJS(data) : data

const isPlainObject = data =>
  data &&
  ((data.constructor && data.constructor.name === 'Object') ||
    data.constructor === undefined)
