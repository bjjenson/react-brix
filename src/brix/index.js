export { BrixConsumer, BrixProvider, BrixContext } from './BrixProvider'
export { default as BoundedSuspense } from './BoundedSuspense'
export { useBrix } from './useBrix'
export { useBrixSetter } from './useBrixSetter'
export { useBrixWorker } from './useBrixWorker'
export { useWorker, getKey } from './useWorker'
