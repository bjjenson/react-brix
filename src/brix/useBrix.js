import { useContext } from 'react'
import { getBrixContext } from './BrixProvider'
import { coerceDataToImmutable } from './helpers'

export const useBrix = (path, notSetValue) => {
  const { state, setState } = useContext(getBrixContext())
  const value = state.getIn(path, notSetValue)

  const update = updatedValue => {
    setState(currState => currState.setIn(path, coerceDataToImmutable(updatedValue)))
  }

  return {
    value,
    set: update,
    cnx: {
      value,
      onChange: e => update(e.target.value),
    },
  }
}
