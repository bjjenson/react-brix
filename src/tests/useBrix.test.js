import { useContext } from 'react'
import { fromJS, Map } from 'immutable'
import { useBrix } from '../brix/useBrix'
import { getBrixContext } from '../brix/BrixProvider'

jest.mock('react')
jest.mock('../brix/BrixProvider')

const setBrixState = jest.fn()
let context, path
beforeEach(() => {
  path = ['user', 'name']
  context = {
    state: fromJS({
      user: {
        name: 'My Name',
      },
    }),
    setBrixState,
  }

  useContext.mockReturnValue(context)
  getBrixContext.mockReturnValue('brix_context')

  setBrixState.mockImplementation(getState => {
    return getState(context.state)
  })
})

test('returns state and setter from brixContext', () => {
  const actual = useBrix(path, Map())
  expect(actual).toMatchSnapshot()
  expect(useContext).toHaveBeenCalledWith('brix_context')
})

test('set updates brix state with value', () => {
  const [, set] = useBrix(path, Map())
  set('New Name')

  expect(setBrixState.mock.calls[0]).toMatchSnapshot()
  expect(setBrixState.mock.results[0]).toMatchSnapshot()
})

test('set updates brix state with function', () => {
  const [, set] = useBrix(path, Map())
  set(name => `updated-${name}`)

  expect(setBrixState.mock.calls[0]).toMatchSnapshot()
  expect(setBrixState.mock.results[0]).toMatchSnapshot()
})
