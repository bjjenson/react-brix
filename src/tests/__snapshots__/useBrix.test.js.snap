// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns state and setter from brixContext 1`] = `
Array [
  "My Name",
  [Function],
]
`;

exports[`set updates brix state 1`] = `
Array [
  [Function],
]
`;

exports[`set updates brix state 2`] = `
Object {
  "isThrow": false,
  "value": Immutable.Map {
    "user": Immutable.Map {
      "name": "New Name",
    },
  },
}
`;
